[NEP-584]: Cross shard bandwidth scheduler

jancionear opened the pull request:
NEP Status (Updated by NEP Moderators)

Status: Approved

SME reviews:
- [X] @shreyan-gupta: https://github.com/near/NEPs/pull/584#issuecomment-2673091851
- [X] @wacban: https://github.com/near/NEPs/pull/584#issuecomment-2671887531


Protocol Work Group voting indications (❔ | 👍 | 👎 ):

* [👍] @birchmd: https://github.com/near/NEPs/pull/584#issuecomment-2672779279 
* [👍] @mfornet: https://github.com/near/NEPs/pull/584#issuecomment-2678965530
* [ ] @mm-near
* [👍] @bowenwang1996: https://github.com/near/NEPs/pull/584#issuecomment-2672933087

walnut-the-cat commented on neps/nep-0584.md:
````diff
@@ -0,0 +1,1421 @@
+}
+```
+
+Additionally, the value that is the closest to `max_receipt_size` is set to to `max_receipt_size`:
````
two 'to'?

walnut-the-cat commented on neps/nep-0584.md:
````diff
@@ -0,0 +1,1421 @@
+}
+```
+
+The values are calculate using a linear interpolation between `base_bandwidth` and
````
'calculated'?

walnut-the-cat commented on neps/nep-0584.md:
````diff
@@ -0,0 +1,1421 @@
+params are:
+
+```rust
+BandwidthSchedulerParams {
+    base_bandwidth: 100000,
+    max_shard_bandwidth: 4500000,
+    max_receipt_size: 4194304,
+    max_allowance: 4500000,
+}
+```
````
nit: clarify how these values are related to the fact that we are practicing the exercise with '4 shards'

jancionear commented:
Matej's comment about the NEP can be found here: https://github.com/near/NEPs/commit/0ff27d178694cb324c13e79cb8d8fe0cc42a7496#commitcomment-151308824

walnut-the-cat commented:
As the moderator,

I want to head-start on the review process for this NEP as the change ins part of 2.5 release.

@jancionear , please comment once you believe this proposal is ready for SME review.

@near/wg-protocol , could you help assign SMEs who can review the proposal?

From engineering perspective, we believe @shreyan-gupta and @wacban are good candidates.

Thank you.

jancionear commented:
> @jancionear , please comment once you believe this proposal is ready for SME review.

I think the NEP is ready for review. I addressed the issues found in Matej's review.

bowenwang1996 commented:
As a working group member, I nominate @shreyan-gupta and @wacban as SME reviewers.

shreyan-gupta reviewed (commented):
Overall great design! I've left a couple of comments.

jancionear commented on neps/nep-0584.md:
```diff
@@ -0,0 +1,1412 @@
+
+### Existing solution
+
+There is already a rudimentary solution in place, added together with stateless validation in
+[NEP-509](https://github.com/near/NEPs/blob/master/neps/nep-0509.md) to limit witness size.
```
It was added after congestion control. The allowed shard from congestion control has to be the same as the shard that is allowed to send more receipts to make sure that there are no liveness issues. If they were different we could have a situation where congestion control allows one shard to send receipts, but it's not the shard that can send large receipts and the large receipts could get stuck.

jancionear commented on neps/nep-0584.md:
```diff
@@ -0,0 +1,1412 @@
+granting `max_single_grant` on one link and dividing it equally between the other links.
+
+There's also a limit which makes sure that `base_bandwidth` stays under 100kB, even when the number
+of shards is low. There are some tests which have a low number of shards, and having a lower base
+bandwidth allows us to fully test the bandwidth scheduler in those tests.
```
Currently this only matters in tests, with current parameters and 6 shards the base bandwidth is ~60kB, and it'll only get smaller as the number of shards increases.

Making `max_shard_bandwidth` larger or `max_single_grant` smaller in the future could make the base bandwidth larger that 100kB, in which case we'll have to reevaluate all the parameters.

shreyan-gupta replied:
Yeah, I guess it shouldn't matter too much given the number of shards can not decrease

jancionear commented on neps/nep-0584.md:
````diff
@@ -0,0 +1,1412 @@
+`max_single_grant`, like this:
+
+```rust
+values[-1] = base_bandwidth
````
> I initially mistook this for the python indexing where value[-1] means the last element of the array.

Damn python strikes again x.x
Added a comment to make it clearer

jancionear commented on neps/nep-0584.md:
````diff
@@ -0,0 +1,1412 @@
+}
+```
+
+The final `BandwidthRequestValues` on current mainnet (6 shards) look like this:
````
Yes that's true.

Around resharding boundary the requests will be slightly wrong for one block - they will be encoded using pre-resharding values, but decoded using post-resharding ones. Ideally we would handle this gracefully, but I decided not to bother, IMO the amount of work needed to do it would be larger than the benefit, everything will go back to normal after one block anyway.

Group sizes could become an issue if they became larger than the distance between two consecutive values, but with more shards the `base_bandwidth` would get smaller, making the values spread out more, so the distance is expected to only increase, it'll never become smaller than group size. Even if group sizes became larger than distance between consecutive values I think nothing would break, it'd just be a bit less efficient.

shreyan-gupta replied:
Makes sense, btw, the values can also change if we change the limit of the `max_single_grant`. Anyway, this shouldn't be a big deal as long as correctness is ensured and this is transient.

jancionear commented on neps/nep-0584.md:
```diff
@@ -0,0 +1,1412 @@
+}
+
+pub struct ReceiptGroupsQueueDataV0 {
+    /// Indices of the receipt groups TrieQueue.
```
Those are not indices of receipts stored in the `BufferedReceipts` queue, they are indices of the `ReceiptGroup`  stored in the `BUFFERED_RECEIPT_GROUPS_QUEUE_ITEM` column.

shreyan-gupta replied:
Makes even more sense to clarify then 😅 I didn't get it in the first go

jancionear commented on neps/nep-0584.md:
```diff
@@ -0,0 +1,1412 @@
+
+#### Allowance
+
+There is a concept of "allowance" - every link (pair of sender and receiver shards) has an
```
Bandwidth scheduler works the same for receipts sent to the same shard. Stateless chunk validation means that receipts from the same shard have to be sent the same way as receipts from other shards, so they must be limited in the same way.

jancionear commented on neps/nep-0584.md:
```diff
@@ -0,0 +1,1412 @@
+- https://github.com/near/nearcore/pull/12728
+- https://github.com/near/nearcore/pull/12747
+
+TODO - am I supposed to copy the code here? I think that a proper "minimal reference implementation"
```
> Maybe including the title of the PR would help a lot.

Great idea, will add

jancionear commented on neps/nep-0584.md:
```diff
@@ -0,0 +1,1412 @@
+   utilization.
+
+
+#### Allowance
```
There is a line about persisting them in the state:

> Link allowances are persisted in the state trie, as they're used to track fairness across multiple heights.

> What's the validity period of an allowance?

Infinite. A link that didn't send anything for a long time should have priority when it finally wants to send something.

> Does an allowance between two shards saturate at some value?

I'll add a paragraph about `max_allowance`, forgot that :+1:

jancionear commented on neps/nep-0584.md:
```diff
@@ -0,0 +1,1412 @@
+is started. The threshold can only be exceeded when a single receipt has size or gas above the group
+threshold.
+
+The size threshold is set to 100kB, the gas threshold is currently infinite.
```
We could do that, although it would be good to add extra checks that group sizes don't get too small as that'd increase the number of trie reads needed to generate the bandwidth requests.

I planned to just put group size parameters next to scheduler params in runtime config, but now I see that I forgot to move them from code to RuntimeConfig x.x

shreyan-gupta replied:
Hmm, sounds good! We can do that the next time we are planning to change the configs.

jancionear commented on neps/nep-0584.md:
```diff
@@ -0,0 +1,1412 @@
+    /// This much bandwidth is granted by default.
+    /// base_bandwidth = (max_shard_bandwidth - max_single_grant) / (num_shards - 1)
+    pub base_bandwidth: Bandwidth,
+    /// The maximum amount of data that a shard can send or receive at a single height.
+    pub max_shard_bandwidth: Bandwidth,
```
No particular reason, it was the easiest to do and there was no need for something more complicated. I guess everything that is sent has to be received, so the amount of data should be similar, assuming equal load.

jancionear commented on neps/nep-0584.md:
```diff
@@ -0,0 +1,1412 @@
+will allow to send more receipts. Just like in `make_request_bitmap_slow`, only using
+`receipt_group_sizes` instead of `buffered_receipt_sizes`.
+
+It's important to note that `size_upper_bound` is less than difference between two consecutive
+values in `BandwidthRequestValues` . Thanks to this the requests are just as good as they would be
+if they were generated directly using individual receipt sizes.
```
That's right, `BandwidthRequest::make_from_receipt_sizes` reads the groups until it reaches a point where this much bandwidth can't be requested with the predefined bandwidth and then it stops. This means that generating a bandwidth request will read at most `max_single_grant / group_size` trie reads, which is about 45 trie reads per outgoing buffer.

jancionear commented on neps/nep-0584.md:
````diff
@@ -0,0 +1,1412 @@
+}
+```
+
+It's important to keep the size of `BandwidthRequest` small because bandwidth requests are included
+in the chunk header, and the chunk header shouldn't be too large.
````
> Just for my information, what would it sum up to for 10/ 50/ 100 shards?

A single bandwidth request takes 6 bytes, with 100 shards the worst case would be 600 bytes per chunk, or 60kB per block. But the worst case is unlikely, usually a shard doesn't do bandwidth requests to all other shards.

> Is there any value in wrapping the bitmap in an Option or does that not affect the number of serialized bytes?

A bandwidth request with zeroed out bitmap wouldn't be included in the list of shard's bandwidth requests, so there's no point to use an `Option`. Bitmap is always nonzero.

> Did you check if the serialized size is what you expect? I could imagine borsh rounding things up to the nearest 32bits for each field, in this case it may be worth to customize the serialization.

I hope there's no rounding, that would be terrible, borsh is supposed to be a one-to-one mapping between structs and serialized data.

jancionear commented on neps/nep-0584.md:
```diff
@@ -0,0 +1,1412 @@
+`base_bandwidth` of receipts, it can just send them out immediately. Actual bandwidth grants based
+on bandwidth request happen after granting the base bandwidth.
+
+On current mainnet (with 6 shards) the base bandwidth is 61_139 (61kB)
```
Added a short note about the relation with number of shards.

`base_bandwidth = base_bandwidth = (max_shard_bandwidth - max_single_grant) / (num_shards - 1)`

`base_bandwidth = (4500000 - 4*1024*1024) / (num_shards - 1)`

For 6 shards:
`(4500000 - 4*1024*1024)/5 = 61139`

For 50 shards:
`(4500000 - 4*1024*1024)/49 = 6238`

For 100 shards:
`(4500000 - 4*1024*1024)/99 = 3087`

So with 100 shards a shard will be able to send at most 3kB of data to another shard without making a request.
This isn't a lot, but it should be enough to send a few receipts.

An additional factor is that as the number of shards increases, the amount of receipts sent on each link should become lower, so a smaller base bandwidth should become less of an issue, it becomes smaller at the same rate as the number of receipts per link becomes smaller.

For larger numbers of shards we could revisit the parameters, for example decreasing `max_single_grant` and `max_receipt_size` to 2MB would greatly increase the budget for base bandwidth, same with increasing `max_shard_bandwidth`.

wacban commented on neps/nep-0584.md:
```diff
@@ -0,0 +1,1425 @@
+
+## Reference Implementation
+
+[This technical section is required for Protocol proposals but optional for other categories. A
+draft implementation should demonstrate a minimal implementation that assists in understanding or
+implementing this proposal. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- Where possible, include a Minimum Viable Interface subsection expressing the required behavior and
+  types in a target programming language. (ie. traits and structs for rust, interfaces and classes
+  for javascript, function signatures and structs for c, etc.)
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+- For protocol changes: A link to a draft PR on nearcore that shows how it can be integrated in the
+  current code. It should at least solve the key technical challenges.
+
+The section should return to the examples given in the previous section, and explain more fully how
+the detailed proposal makes those examples work.]
```
I think you can remove that

wacban reviewed (approved):
LGTM, thanks!

flmel commented:
@near/wg-protocol Can you please fully read this NEP and comment in the thread if you are leaning towards approving or rejecting it? Please make sure to include your rationale and any feedback that you have for the author.

wacban commented:
As a subject matter expert, I approve this NEP.

birchmd commented:
As a working group member I lean towards approving this proposal. It is detailed, well-written and solves an important problem; thanks for all your work on it!

My only comment is that it would be nice to present some Byzantine analysis of the bandwidth scheduling sub-protocol. What happens if someone maliciously modifies their node to send malformed or otherwise incorrect bandwidth messages? It doesn't look to me at first glance like there is a serious issue here, but I still wanted to raise it for the designers to take a look at.

bowenwang1996 commented:
As a working group member I lean towards approving this proposal

shreyan-gupta commented on neps/nep-0584.md:
````diff
@@ -0,0 +1,1412 @@
+```rust
+let mut sanity_check_bytes = Vec::new();
+sanity_check_bytes.extend_from_slice(scheduler_state.sanity_check_hash.as_ref());
+sanity_check_bytes.extend_from_slice(CryptoHash::hash_borsh(&all_shards).as_ref());
````
Yeah, maybe later we can revisit what is being stored in `sanity_check_bytes`. The current info may not be too useful to us in case there are some inconsistencies related to the allowance calculation across shards/nodes.

Perhaps there may be a cleverer way of getting a digest for the state.

shreyan-gupta commented:
As a subject matter expert, I am leaning towards approving this NEP.

Thanks Jan for working on this problem. The proposal is very well written. It's a great step towards a problem that we aniticiapte to hit quite quickly as our blockchain usage and number of shards grow.

While there are specifics we can improve, and revisit, and perhaps redesign over time, they are of minor importance as of the current status of NEAR blockchain. Examples include revisiting the limit calculations, `BandwidthRequestValues`, `sanity_check_bytes`, integration with congestion control etc.

The bandwidth scheudler is designed in a way that it is easily extendible and can be moulded in any direction we may see fit based on the future requirements.

jancionear commented:
> My only comment is that it would be nice to present some Byzantine analysis of the bandwidth scheduling sub-protocol. What happens if someone maliciously modifies their node to send malformed or otherwise incorrect bandwidth messages? It doesn't look to me at first glance like there is a serious issue here, but I still wanted to raise it for the designers to take a look at.

Bandwidth scheduler is run during chunk application and its results (the produced bandwidth requests) are stored in the chunk header, along with other things produced when applying a chunk. Chunk validators apply the chunk and verify that produced data matches the data in chunk header, they will not endorse a chunk if it doesn't match. The data from previous chunk header (like previous bandwidth requests) can be trusted because the previous chunk headers were endorsed by chunk validators at the previous height. The logic is pretty much identical to `CongestionInfo`.

You're right, I forgot to describe that in the NEP. I can add a small section, but I'm not sure if it's ok to modify the NEP at this stage?

flmel commented:
> > My only comment is that it would be nice to present some Byzantine analysis of the bandwidth scheduling sub-protocol. What happens if someone maliciously modifies their node to send malformed or otherwise incorrect bandwidth messages? It doesn't look to me at first glance like there is a serious issue here, but I still wanted to raise it for the designers to take a look at.
> 
> Bandwidth scheduler is run during chunk application and its results (the produced bandwidth requests) are stored in the chunk header, along with other things produced when applying a chunk. Chunk validators apply the chunk and verify that produced data matches the data in chunk header, they will not endorse a chunk if it doesn't match. The data from previous chunk header (like previous bandwidth requests) can be trusted because the previous chunk headers were endorsed by chunk validators at the previous height. The logic is pretty much identical to `CongestionInfo`.
> 
> I forgot to describe that in the NEP. I can add a small section, but I'm not sure if it's ok to modify the NEP at this stage?

If it is something implemented but not described I believe it can be added taking into consideration Workgroup member recommendation.

mfornet commented:
As a working group member I lean towards approving this proposal

jancionear commented:
Added a section about byzantine fault tolerance
